presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockEncoding.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.type.TypeManager;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+
+public class ArrayBlockEncoding
+        implements BlockEncoding
+{
+    public static final BlockEncodingFactory<ArrayBlockEncoding> FACTORY = new ArrayBlockEncodingFactory();
+    private static final String NAME = "ARRAY";
+
+    private final BlockEncoding valueBlockEncoding;
+
+    public ArrayBlockEncoding(BlockEncoding valueBlockEncoding)
+    {
+        this.valueBlockEncoding = valueBlockEncoding;
+    }
+
+    @Override
+    public String getName()
+    {
+        return NAME;
+    }
+
+    @Override
+    public void writeBlock(SliceOutput sliceOutput, Block block)
+    {
+        AbstractArrayBlock arrayBlock = (AbstractArrayBlock) block;
+
+        valueBlockEncoding.writeBlock(sliceOutput, arrayBlock.getValues());
+        int positionCount = arrayBlock.getPositionCount();
+        sliceOutput.appendInt(positionCount);
+        sliceOutput.appendBytes(arrayBlock.getOffsets().slice(0, (positionCount + 1) * 4));
```

Comment:
is it necessary to have the `.slice()` call here? That looks like it should be a no-op